File: src/pensum/web/static/primitives/ten-frame.js
/* One or two ten-frames.
 *
 * State: `{frames: [[cells], [cells]]}`, the filled cells of each frame
 * numbered in reading order 0..9, the same shape the server grades
 * (`pensum.items.primitives.ten_frame`). Zones `f0` and `f1` are the frames;
 * a zone's index is its cell.
 *
 * Tap an empty cell to put a dot there; tap a dot to pick it up and again to
 * take it away; drag a dot to another cell. With a keyboard the frame itself
 * takes focus: the arrow keys move a cursor over the cells and Space or Enter
 * puts a dot down or takes one away.
 */
(function () {
  "use strict";

  function tenFrameParse(json, limits) {
    var raw;
    try {
      raw = JSON.parse(json);
    } catch (error) {
      return null;
    }
    if (!raw || !Array.isArray(raw.frames) || raw.frames.length !== limits.frames) {
      return null;
    }
    var frames = [];
    for (var f = 0; f < raw.frames.length; f++) {
      var cells = raw.frames[f];
      if (!Array.isArray(cells)) return null;
      var seen = {};
      for (var c = 0; c < cells.length; c++) {
        var cell = cells[c];
        if (typeof cell !== "number" || cell !== Math.floor(cell) || cell < 0 || cell >= limits.cells) {
          return null;
        }
        if (seen[cell]) return null;
        seen[cell] = true;
      }
      frames.push(cells.slice().sort(function (a, b) { return a - b; }));
    }
    return { frames: frames };
  }

  function tenFrameHas(state, frame, cell) {
    return state.frames[frame] !== undefined && state.frames[frame].indexOf(cell) >= 0;
  }

  function tenFrameFrame(zone) {
    var match = /^f(\d+)$/.exec(zone || "");
    return match ? +match[1] : -1;
  }

  function tenFrameSet(state, frame, cell, on) {
    var frames = state.frames.map(function (cells) { return cells.slice(); });
    if (on) {
      frames[frame].push(cell);
      frames[frame].sort(function (a, b) { return a - b; });
    } else {
      frames[frame] = frames[frame].filter(function (c) { return c !== cell; });
    }
    return { frames: frames };
  }

  function tenFrameValid(state, frame, cell, limits) {
    return frame >= 0 && frame < state.frames.length && cell >= 0 && cell < limits.cells;
  }

  function tenFrameApply(state, action, limits) {
    var frame;
    var f;
    var c;
    if (action.type === "add") {
      if (action.zone !== undefined && action.index !== undefined && action.index >= 0) {
        frame = tenFrameFrame(action.zone);
        if (!tenFrameValid(state, frame, action.index, limits)) return null;
        if (tenFrameHas(state, frame, action.index)) return null;
        return tenFrameSet(state, frame, action.index, true);
      }
      /* The button: the first empty cell in reading order. */
      for (f = 0; f < state.frames.length; f++) {
        for (c = 0; c < limits.cells; c++) {
          if (!tenFrameHas(state, f, c)) return tenFrameSet(state, f, c, true);
        }
      }
      return null;
    }
    if (action.type === "remove" || action.type === "act") {
      if (action.type === "act") {
        frame = tenFrameFrame(action.zone);
        if (!tenFrameValid(state, frame, action.index, limits)) return null;
        if (!tenFrameHas(state, frame, action.index)) return null;
        return tenFrameSet(state, frame, action.index, false);
      }
      /* The button: the last filled cell in reading order. */
      for (f = state.frames.length - 1; f >= 0; f--) {
        var cells = state.frames[f];
        if (cells.length) return tenFrameSet(state, f, cells[cells.length - 1], false);
      }
      return null;
    }
    if (action.type === "toggle") {
      frame = tenFrameFrame(action.zone);
      if (!tenFrameValid(state, frame, action.index, limits)) return null;
      return tenFrameSet(state, frame, action.index, !tenFrameHas(state, frame, action.index));
    }
    if (action.type === "move") {
      var from = tenFrameFrame(action.from);
      var to = tenFrameFrame(action.to);
      if (!tenFrameValid(state, from, action.fromIndex, limits)) return null;
      if (!tenFrameValid(state, to, action.toIndex, limits)) return null;
      if (!tenFrameHas(state, from, action.fromIndex) || tenFrameHas(state, to, action.toIndex)) {
        return null;
      }
      return tenFrameSet(tenFrameSet(state, from, action.fromIndex, false), to, action.toIndex, true);
    }
    return null;
  }

  function tenFrameShown(state, piece) {
    return tenFrameHas(state, tenFrameFrame(piece.zone), piece.index);
  }

  function tenFrameDescribe(state, say) {
    var total = state.frames.reduce(function (sum, cells) { return sum + cells.length; }, 0);
    return window.PensumActivity.plural(say, "made", total);
  }

  /* Where the keyboard cursor goes from `at` on a key, across `frames`
   * frames of two rows of `perRow`. Up and down move a row, and cross from
   * one frame into the next. -1 for a key that does not move it. */
  function tenFrameStep(at, key, frames, perRow) {
    var last = frames * 2 * perRow - 1;
    var next = -1;
    if (key === "ArrowRight") next = at + 1;
    else if (key === "ArrowLeft") next = at - 1;
    else if (key === "ArrowDown") next = at + perRow;
    else if (key === "ArrowUp") next = at - perRow;
    else if (key === "Home") next = 0;
    else if (key === "End") next = last;
    else return -1;
    return Math.max(0, Math.min(last, next));
  }

  function tenFrameBind(root, api) {
    var board = api.board;
    var zones = board.querySelectorAll("[data-drop]");
    var cursor = 0;
    board.setAttribute("tabindex", "0");
    board.setAttribute("role", "application");

    function show() {
      for (var i = 0; i < zones.length; i++) {
        zones[i].classList.toggle("is-cursor", i === cursor);
      }
      var frame = Math.floor(cursor / api.limits.cells);
      var cell = cursor % api.limits.cells;
      api.announce(
        window.PensumActivity.fill(api.say.cell, {
          frame: frame + 1,
          cell: cell + 1,
          state: tenFrameHas(api.state(), frame, cell) ? api.say.full : api.say.empty,
        })
      );
    }

    board.addEventListener("focus", show);
    board.addEventListener("blur", function () {
      for (var i = 0; i < zones.length; i++) zones[i].classList.remove("is-cursor");
    });
    board.addEventListener("keydown", function (event) {
      if (event.key === " " || event.key === "Enter") {
        event.preventDefault();
        var frame = Math.floor(cursor / api.limits.cells);
        api.dispatch({ type: "toggle", zone: "f" + frame, index: cursor % api.limits.cells });
        show();
        return;
      }
      var next = tenFrameStep(cursor, event.key, api.limits.frames, api.limits.perRow);
      if (next >= 0) {
        event.preventDefault();
        cursor = next;
        show();
      }
    });
  }

  if (window.PensumActivity) {
    window.PensumActivity.register("ten_frame", {
      parse: tenFrameParse,
      apply: tenFrameApply,
      shown: tenFrameShown,
      describe: tenFrameDescribe,
      bind: tenFrameBind,
    });
  }
})();

File: src/pensum/items/primitives/ten_frame.py
"""One or two ten-frames: two rows of five cells, a dot in some of them.

A ten-frame makes numbers to twenty visible as "five and a bit" and "ten and a
bit", which is what lets a child see 8 without counting it and see that 8 needs
2 more to make ten. The frame is the most-used manipulative in early number for
that reason.

The pupil taps an empty cell to put a dot in it, taps a dot to pick it up and
taps it again to take it away, or drags a dot to another cell. With a keyboard,
the arrow keys move a cursor over the cells and Space puts a dot down or takes
it away; the add and remove buttons do the same in reading order.

Graded on the count, and optionally on the fill pattern: `order: in_order`
asks for the conventional fill, top row left to right before the bottom row and
the first frame before the second, which is the pattern that makes "ten and
three" readable at a glance.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from pensum.i18n import translate
from pensum.items.figures import Path
from pensum.items.primitives.base import (
    ActivityConfig,
    Board,
    Layer,
    Piece,
    Zone,
    plural,
    say_keys,
)

CELLS = 10
PER_ROW = 5
CELL = 30.0
RADIUS = 10.0
MARGIN = 12.0
FRAME_GAP = 14.0


class TenFrameState(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")

    # Which cells hold a dot, per frame, numbered in reading order 0..9.
    frames: tuple[tuple[int, ...], ...]


class TenFrameActivity(ActivityConfig):
    """Show `target` dots on the frames."""

    FALLBACK_KEY = "activity.ten_frame.ask_count"
    State = TenFrameState

    kind: Literal["ten_frame"] = "ten_frame"
    frames: int = Field(default=1, ge=1, le=2)
    target: int = Field(ge=0, le=2 * CELLS)
    # Dots already there, filled in reading order: the "8" in "make 8 into 10".
    start: int = Field(default=0, ge=0, le=2 * CELLS)
    order: Literal["any", "in_order"] = "any"

    @model_validator(mode="after")
    def _check(self) -> TenFrameActivity:
        room = self.frames * CELLS
        if self.target > room or self.start > room:
            raise ValueError(f"{self.frames} frame(s) hold {room} dots at most")
        return self

    def _adding(self) -> bool:
        return 0 < self.start < self.target

    def fallback_key(self) -> str:
        return "activity.ten_frame.ask_add" if self._adding() else self.FALLBACK_KEY

    def _first(self, count: int) -> TenFrameState:
        """The first `count` cells in reading order, across the frames."""
        return TenFrameState(
            frames=tuple(
                tuple(range(max(0, min(CELLS, count - frame * CELLS))))
                for frame in range(self.frames)
            )
        )

    def initial(self) -> TenFrameState:
        return self._first(self.start)

    def solution(self) -> TenFrameState:
        return self._first(self.target)

    def admits(self, state: TenFrameState) -> bool:
        if len(state.frames) != self.frames:
            return False
        for cells in state.frames:
            if len(set(cells)) != len(cells) or any(not 0 <= c < CELLS for c in cells):
                return False
        return True

    def grade_state(self, state: TenFrameState) -> bool:
        if _count(state) != self.target:
            return False
        if self.order == "any":
            return True
        return _filled(state) == _filled(self.solution())

    def default_fallback(self) -> float:
        return float(self.target - self.start) if self._adding() else float(self.target)

    def from_number(self, value: float) -> TenFrameState | None:
        if not float(value).is_integer() or not 0 <= value <= self.frames * CELLS:
            return None
        if self._adding():
            # The typed number answered "how many more", so the board shows
            # the frame with that many added to what was there.
            total = self.start + int(value)
            return self._first(total) if total <= self.frames * CELLS else None
        return self._first(int(value))

    def describe(self, state: TenFrameState, locale: str) -> str:
        text = plural(locale, "activity.ten_frame.made", _count(state))
        if (
            self.order == "in_order"
            and _count(state) == self.target
            and _filled(state) != _filled(self.solution())
        ):
            text += ", " + translate(locale, "activity.ten_frame.out_of_order")
        return text

    def limits(self) -> dict[str, object]:
        return {"frames": self.frames, "cells": CELLS, "perRow": PER_ROW}

    def say(self, locale: str) -> dict[str, str]:
        return say_keys(
            locale,
            "activity.ten_frame.made",
            "activity.ten_frame.made_one",
            "activity.ten_frame.cell",
            "activity.ten_frame.full",
            "activity.ten_frame.empty",
        )

    def board(self, state: TenFrameState, locale: str) -> Board:
        width = 2 * MARGIN + PER_ROW * CELL
        paths: list[Path] = []
        zones: list[Zone] = []
        pieces: list[Piece] = []
        for frame in range(self.frames):
            top = MARGIN + frame * (2 * CELL + FRAME_GAP)
            filled = set(state.frames[frame]) if frame < len(state.frames) else set()
            # The frame: an outer box, a line between the rows, and a line
            # between each cell. Drawn as strokes, not as the zones, so the
            # frame is there without a script and the zones stay invisible.
            right, bottom = MARGIN + PER_ROW * CELL, top + 2 * CELL
            paths.append(
                Path(
                    f"M{MARGIN:.2f},{top:.2f}L{right:.2f},{top:.2f}L{right:.2f},{bottom:.2f}"
                    f"L{MARGIN:.2f},{bottom:.2f}Z",
                    "outline",
                )
            )
            paths.append(Path(f"M{MARGIN:.2f},{top + CELL:.2f}L{right:.2f},{top + CELL:.2f}"))
            for column in range(1, PER_ROW):
                x = MARGIN + column * CELL
                paths.append(Path(f"M{x:.2f},{top:.2f}L{x:.2f},{bottom:.2f}"))
            for cell in range(CELLS):
                row, column = divmod(cell, PER_ROW)
                x, y = MARGIN + column * CELL, top + row * CELL
                zones.append(Zone(f"f{frame}", x, y, CELL, CELL, index=cell, role="zone cell"))
                inset = CELL / 2 - RADIUS
                pieces.append(
                    Piece(
                        f"f{frame}",
                        cell,
                        "counter",
                        x + inset,
                        y + inset,
                        2 * RADIUS,
                        2 * RADIUS,
                        cell in filled,
                        row=row,
                        col=column,
                    )
                )
        height = MARGIN * 2 + self.frames * 2 * CELL + (self.frames - 1) * FRAME_GAP
        return Board(
            width,
            height,
            self.alt.get(locale),
            (Layer("base", tuple(paths), tuple(zones), tuple(pieces)),),
        )


def _count(state: TenFrameState) -> int:
    return sum(len(cells) for cells in state.frames)


def _filled(state: TenFrameState) -> tuple[frozenset[int], ...]:
    return tuple(frozenset(cells) for cells in state.frames)
